### DISCUSSION TITLE
Performance issues

### DISCUSSION THREAD

**Discussion by maximkrouk** [pointfreeco/swift-composable-architecture#585] [General]
Hi, I'm using RAS-TCA and I'm facing noticeable performance issues. The app has about 40 modules and a large number of actions, the state is pretty large too. Partially issues may not be connected with TCA itself, but some of them may relate.

I created the thread for discussing ideas on how to improve the performance and to keep the history of the performance improvement process reachable for myself and other developers.

[*The App*](https://apps.apple.com/app/apple-store/id1532032195)

### Issues
- High CPU usage
  - Laggy scroll
  - Device temperature
  
 ### Possible reasons
 - AVPlayer
 - [CasePaths](https://github.com/pointfreeco/swift-case-paths/discussions/25)
 - Amount of reducers
 - ReactiveSwift (_[Link 1](https://github.com/trading-point/reactiveswift-composable-architecture/pull/41), [Link 2](https://github.com/ReactiveCocoa/ReactiveSwift/pull/797)_)

**Comment by maximkrouk** [comment]
The first step is AVPlayer improvements, we didn't fix it completely, but extracting most of the player logic to a background thread and moving player setup from `cellWillAppear` to `scrollViewDidDecelerate` noticeable improved scroll performance, but anyway it's not perfect yet.

AVPlayer notes can be found [here](https://www.craft.do/s/2Ic8O8TrFiK8L2)

Btw player is not an environmental dependency, but a shared instance for the app, player states are stored in feed elements' states and FeedItemStore for each FeedItem is scoped to playerStore and set to the player on `scrollViewDidDecelerate`

**Comment by maximkrouk** [comment]
Tomorrow I'll try to [generate custom CasePaths for all of our Actions](https://github.com/pointfreeco/swift-case-paths/discussions/25#discussioncomment-616055), and reply here if it fixes any issues or not.

**Reply by stephencelis** [reply]
Let us know if this is the culprit. We're exploring improving the reflection mechanism soon, but it'd be helpful to have some data points.

**Reply by maximkrouk** [reply]
Couldn't generate it yet, seems like I need a little more time to implement it.

**Reply by maximkrouk** [reply]
Well, basic script is done, I'll fix a few issues, cleanup the API and publish it this week.
Also, I integrated generated CasePaths into a project, seems like it partially improved app performance, but I'll show the update to my team and give you a feedback tomorrow.

**Reply by maximkrouk** [reply]
Well, the app works a bit better on iPhone 11, but 5s/6 are still laggy

**Comment by maximkrouk** [comment]
Also, my colleague proposed to use a serial background scheduler for all of the application logic and only observe UI logic on the main thread (and also send actions on the same background scheduler)
@mbrandonw, @stephencelis, @mluisbrown, what do you think about this approach?

**Reply by mluisbrown** [reply]
TCA is designed so that all interactions with the store must be done on the same thread. See the [FAQ](https://github.com/pointfreeco/swift-composable-architecture#faq) for more details:

> All interactions with an instance of Store (including all of its scopes and derived ViewStores) must be done on the same thread. If the store is powering a SwiftUI or UIKit view then, all interactions must be done on the main thread.

`Store` is not thread safe. I'm not sure how easy it would be to split application logic and UI updates into separate threads.

**Reply by maximkrouk** [reply]
We use a ComposableCore object that wraps an optional store and viewStore and used to get state and send actions, so it's pretty easy to migrate to background store, we just should change `core.sendAsync(Action, on: DispatchQueue? = .main)` method to `core.sendAsync(Action, on: DispatchQueue? = .app)`, change `mainQueue: .main` in AppEnvironment to `mainQueue: .app` and hop to UIScheduler in UI-bindings, but I'm not sure if we get some other problems if we do that.
If generated CasePaths won't help, maybe we'll just try and find out if it works on practice 🙂

**Reply by AlexisQapa** [reply]
IIRC there was a previous discussion doing so for some kind of headless TCA. AFAIK there is no issue as long as you don't use features like animations which should be run on the main queue.

Thought I think your reducer shouldn't be that slow maybe you are doing too much like sorting big arrays which could be done on a background queue as an effect and keep your stone one the main queue.

**Reply by maximkrouk** [reply]
Well, it's the next step, so we'll try to move our reducers to a background scheduler and reply here if it works fine

**Reply by maximkrouk** [reply]
We moved appLogic to a background thread and it increased app performance a lot (even iPhone 5S works almost smoothly).

But now UI is significantly faster than logic, so scroll events, for example, are processed some time after scroll and new actions can be processed only after all previous ones are processed (it can take more then 30s to process them 😠)

We'll try to investigate it a bit more and maybe throttle some actions or use ComposableCollectionView with a local (non-scoped) store. But I'm not sure, that this behavior (fast ui, slow logic) won't lead to some new issues 🤔

**Reply by shawnkoh** [reply]
> We moved appLogic to a background thread and it increased app performance a lot (even iPhone 5S works almost smoothly).
> 
> But now UI is significantly faster than logic, so scroll events, for example, are processed some time after scroll and new actions can be processed only after all previous ones are processed (it can take more then 30s to process them 😠)
> 
> We'll try to investigate it a bit more and maybe throttle some actions or use ComposableCollectionView with a local (non-scoped) store. But I'm not sure, that this behavior (fast ui, slow logic) won't lead to some new issues 🤔

Hi! Do you mind sharing how this has worked out for you?

**Reply by maximkrouk** [reply]
@mluisbrown, we experimented a bit and moved our app logic to background thread, and achieved smooth scrolling, but the first issue (not related to rac-tca, since we had to send actions asynchronously anyway to avoid deadlocks) is that depending on implementation you may lose some sync-ability, not critical in most of scenarios, but just a thing to note. The second and the one that made us revert changes is that even if UI is not affected by store performance bottlenecks, stacking actions could add up to multiple seconds and any interaction is delayed by the time of executing already-sent actions.

There were multiple things that drained the performance afaiu. The first is sending actions too frequently, it can be partially solved by extracting critical parts of the global state or throttling actions from UI layer. The second is probably using StorePublisher's dynamicMemberLookup for triggering collectionView updates which causes more complex equations than just `publisher[dynamicMember: \.items.count]`.

So the conclusion is:
- Switching to background stores may improve UI performance, but leads to another problems, so we decided to keep stores on the main thread
- Try reduce the amount of actions in critical parts even if it can be a little bit inconsistent in terms of architecture
- Some updates can be less time consuming if removeDuplicates is used more frequently, but on a simpler substate, but it's related more to UIKit, than SwiftUI, since in SwiftUI this mechanism allows you to avoid redundant View updates.

UPD:
I replied to mluisbrown (he wrote the first, not the last comment 😅) instead of replying to @shawnkoh accidentally 🌚

**Comment by ferologics** [comment]
From my experience the performance hits the fan with TCA whenever you send many actions through multiple layers of pullback reducers. I.e. calling `onAppear/Disappear` from a scoped `ViewStore` can be super expensive when it has to bubble up and down a `app > chat > message` reducer chain. For this reason we had to postpone the dynamic content prefetching implementation (that depended on the `onAppear/Disappear` actions of each list item). Sending that many actions via our reducer chain (with just 3 pullbacks) was a big performance bottleneck and made the list very choppy/unusable.

<details><summary>Screenshot of trace</summary>

<img width="1904" alt="image" src="https://user-images.githubusercontent.com/5576161/121349269-9a082800-c929-11eb-8633-6b2c43c6f6b5.png">

</details>

[TCA_prefetching_trace.trace.zip](https://github.com/pointfreeco/swift-composable-architecture/files/6623489/TCA_prefetching_trace.trace.zip)

**Reply by siejkowski** [reply]
I've been experiencing the same problems and had the same realisation. Sending the action through a number of layers of pullback reducers is very expensive, performance-wise.

The workaround I've landed on was to introduce a separate root Store that was designed to perform only the actions in the critical path. It is not part of the AppStore and therefore it requires some careful managements to keep in sync with the main AppState.

It's not a solution I'm happy with, but at least I can smoothly scroll through LazyGrid with thousands of asynchronously loaded images.

More info and also the very similar discussion from a few months back: https://github.com/pointfreeco/swift-composable-architecture/discussions/382#discussioncomment-357248

**Reply by maximkrouk** [reply]
Probably some of the pullback issues might be caused by runtime reflection in CasePaths, but it's probably only a part of the problem and CasePaths generation didn't fix performance enough for EduDo on iPhone 6/6s
But it was mentioned in a CasePaths performance discussion, that it solved an issue for another app...
Hopefully we'll test a background store this week (brutally implemented migration for our feed caused to crashes, probably the developer who tried it forgot to change environmental main queue or move ui bindings to the main thread, but the app became usable at iPhone 5S (as he said 🌚))

So I hope it will work, otherwise maybe we'll take a look at extracting some critical paths to another store or extracing some logic into background clients if possible...

**Reply by maximkrouk** [reply]
Btw we have a ton of actions 😅

Composable player 😎 (sends 2 or 3 actions every 0.15s (max pullback depth is currently 10) 🤡)
Composable CollectionView 😎(sends actions on every `scrollViewDidScroll` call 🤡)
ComposableViewController 😎 (sends actions on every view lifecycle event (except of `viewDidLayoutSubviews`) 🤡)

**Reply by siejkowski** [reply]
That's quite a lot :D Reducing the number of actions helps, I got some gains by caching some information in the store and not sending actions when possible.
I've too hand-written the CasePaths, it helped a little but not enough to achieve the smooth scrolling.

And I've also worked on app that uses the media player 😄 The performance hit by sending the action every 0,5s is truly awful! I've not yet optimised it, but it's mind boggling to see the CPU usage jump by 10% when you start sending actions with the player position.

I'm really looking forward to seeing the results of your investigation into background stores. I've done only some small experiments there and never really got it right before moving to another project.

**Reply by mbrandonw** [reply]
We have some changes coming to TCA soon that will improve the performance related to `.scope` and view store observations, and we are in the early stages of experimenting with a more performant form of case path reflection. We'd be really curious to see how your instrumentation data changes after using those changes. Will have an update soon!

**Reply by ferologics** [reply]
@mbrandonw 👌 it should be as easy as re-enabling the currently disabled prefetching implementation -- I can also help benchmark some pre-release version/branch. The code is frozen ATM so any performance improvements should be visible.

**Reply by mrackwitz** [reply]
@mbrandonw How are these changes coming along? Is there anything I could test drive or support with?

**Reply by stephencelis** [reply]
@mrackwitz We just pushed a `performance-tweaks` branch that aggregates a few upcoming changes:

- Some cleanup to `store.send` to coalesce multiple mutations for synchronous effects into a single mutation, which means fewer emissions of state to downstream publishers
- Some cleanup to `store.scope` to avoid extra evaluations of the `toLocalState` transform function
- Some cleanup to `ViewStore` to avoid extra equality checks
- Some enhancements to Case Path's `extract` function that should boost extraction performance by over 2x on successful extraction, and up to 100x faster on failed extraction.

The first 3 updates we've been testing and plan to release on Monday, so any early feedback would be helpful. The `CasePath` change is a more recent endeavor so we'll be giving it another week.

**Reply by mrackwitz** [reply]
@stephencelis Thanks for sharing. This looks very promising. I'll try to have a look into this soon and will report feedback.

**Comment by maximkrouk** [comment]
Btw we migrated to the original TCA this weak, so we'll be able to check performance tweaks branch soon 😎
Also, I had no time to improve our CasePath generation script, but it's already available on https/GitHub.com/edudo-inc/spmgen casepath branch 🙂
